### PR TITLE
Add Metric abstraction so metrics can be added to pipeline

### PR DESCRIPTION
**Pull request by HoagyC**
Tries to create a flexible metrics abstraction that ensures that additional metrics don't clog the train loop and which allows users of the library to pass their own metrics into the train loop without having to edit the library itself.

Changes:

- Implemented a new base Metric class.
- Converted existing metrics to subclasses of Metric.
- Introduced MetricArgs TypedDict for passing generic metric args to 
- Integrated these changes into the training loop.
- New metrics can be added by users with the `additional_metrics` argument to `pipeline`.

### CODE REVIEW

File: sparse_autoencoder/train/metrics/metric_class.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    learned_activations: Float[Tensor, "batch learned_activations"]
+    reconstructed_activations: Float[Tensor, "batch input_activations"]
+
+
+class Metric(ABC):
+    """Base class for metrics."""
+
+    @abstractmethod
+    def compute_and_log(self, args: MetricArgs) -> None:
+        """Compute the metric value."""
```
Comment (alan-cooney) [inline sparse_autoencoder/train/metrics/metric_class.py:30]: Definitely think it makes sense to have a base Metric class - but I imagine there are 3 types of abstract class that extend from this - metrics during data generation, metrics during training and validation metrics (e.g. total loss on the source model with and without the sae patched in)

File: sparse_autoencoder/train/metrics/metric_class.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    """Class to hold arguments to metrics, contains everything needed to compute any metric."""
+
+    step: int
+    batch: Float[Tensor, "batch input_activations"]
+    reconstruction_loss_mse: Float[Tensor, " item"]
+    l1_loss: Float[Tensor, " item"]
+    autoencoder: SparseAutoencoder
+    optimizer: Optimizer
+    learned_activations: Float[Tensor, "batch learned_activations"]
+    reconstructed_activations: Float[Tensor, "batch input_activations"]
```
Comment (alan-cooney) [inline sparse_autoencoder/train/metrics/metric_class.py:22]: Great to have a typed dict for this, as it means we can extend the number of parameters passed without breaking peoples existing loss functions. I think we should add this part in (noting I've got a base metric class I'll probably have merged in by the time you read this so I think we can just update it)

However we shouldn't be passing the autoencoder/optimizer in as this couples these classes with the loss classes and makes it hard to reason about the very different objects. Instead we're using a mediator pattern (aka a pipeline) so the mediator should get any data we need from these (e.g. the current weights from the model) and pass them in directly to the loss functions.

File: sparse_autoencoder/train/metrics/capacity.py [hunk 3]
```diff
@@ -62,3 +64,13 @@ def wandb_capacities_histogram(
 
     bins, values = histogram(numpy_capacities, bins=20, range=(0, 1))
     return wandb.Histogram(np_histogram=(bins, values))
+
+
+class CapacityMetric(Metric):
+    """Capacity metric."""
```
Comment (alan-cooney) [inline sparse_autoencoder/train/metrics/capacity.py:70]: Looks good

File: sparse_autoencoder/train/metrics/metric_class.py (hunk 1 above)
Comment (alan-cooney) [reply sparse_autoencoder/train/metrics/metric_class.py:30]: Should also flag that we ideally don't want the metrics class to store state, unless absolutely necessary. Where possible we can just log things to WandB and then calculate metrics that track history there

**Review by alan-cooney — CHANGES_REQUESTED** [review]
Hey thanks for working on this - as I had some code started (not much) I'll merge that in and then we can add some of the things here into it as well.

Btw https://docs.kidger.site/equinox/pattern/ is a good (short) read that explains the general idea behind abstract/final patterns.

**Comment by alan-cooney** [comment]
Closing for now as the metrics are being worked on separately
